piq/pieapp.py

```diff
@@ -0,0 +1,205 @@
+        stride: Step between cropped patches. Smaller values lead to better quality,
+            but cause higher memory consumption. Default: 27 (`sparse` sampling in original implementation)
+        replace_pooling: Flag to replace MaxPooling layer with AveragePooling. See [3] for details.
+
+    References:
+        .. [1] Ekta Prashnani, Hong Cai, Yasamin Mostofi, Pradeep Sen
+            (2018). PieAPP: Perceptual Image-Error Assessment through Pairwise Preference
+            https://arxiv.org/abs/1806.02067
+        .. [2] https://github.com/prashnani/PerceptualImageError
+        .. [3] Gatys, Leon and Ecker, Alexander and Bethge, Matthias
+            (2016). A Neural Algorithm of Artistic Style}
+            Association for Research in Vision and Ophthalmology (ARVO)
+            https://arxiv.org/abs/1508.06576
+    """
+    # TODO(zakajd) 10/12/2020: Load weights to release and change this link
+    _weights_url = "https://web.ece.ucsb.edu/~ekta/projects/PieAPPv0.1/weights/PieAPPv0.1.pth"
+
+    def __init__(
+        self,
+        reduction: str = "mean",
+        data_range: Union[int, float] = 1.0,
+        stride: int = 27,
+        replace_pooling: bool = False
+    ) -> None:
+        super().__init__()
+        
+        # Load weights and initialize model
+        weights = torch.hub.load_state_dict_from_url(self._weights_url, progress=False)
+        # Fix small bug in original weights
+        weights['ref_score_subtract.weight'] = weights['ref_score_subtract.weight'].unsqueeze(1)
+        self.model = PieAPPModel(features=64)
+        self.model.load_state_dict(weights)
+
+        if replace_pooling:
+            self.model = self.replace_pooling(self.model)
+
+        # Disable gradients
+        for param in self.model.parameters():
+            param.requires_grad_(False)
+
+        self.data_range = data_range
+        self.reduction = reduction
+        self.stride = stride
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of PieAPP  between feature representations of prediction and target tensors.
+        Args:
+            prediction: Tensor with shape (H, W), (C, H, W) or (N, C, H, W).
+            target: Tensor with shape (H, W), (C, H, W) or (N, C, H, W).
+        """
+        _validate_input(input_tensors=(prediction, target), allow_5d=False, allow_negative=True)
+        prediction, target = _adjust_dimensions(input_tensors=(prediction, target))
+
+        N, C, _, _ = prediction.shape
+
+        self.model.to(device=prediction.device)
+        prediction_features, prediction_weights = self.get_features(prediction)
+        target_features, target_weights = self.get_features(target)
+
+        distances, weights = self.model.compute_difference(
+            target_features - prediction_features,
+            target_weights - prediction_weights
+        )
+
+        # Shape (N, NUM_PATCHES)
```

Comment:
What is `NUM_PATCHES`? I don't see this variable in the module

Reply:
Removed this comment. 
Number of patches depend on input image size, crop size and stride, it's not some fixed value.